src/config/mongoose_config_parser_toml.erl

```diff
@@ -292,31 +296,76 @@ node_to_string({host, _}) -> [];
 node_to_string(Node) -> [binary_to_list(Node)].
 
 -spec item_key(path(), toml_value()) -> {host, jid:server()} | item.
+item_key([<<"host_config">>], #{<<"host_type">> := Host}) -> {host, Host};
 item_key([<<"host_config">>], #{<<"host">> := Host}) -> {host, Host};
 item_key(_, _) -> item.
 
 %% Processing of the parsed options
 
--spec get_hosts(config_list()) -> [jid:server()].
-get_hosts(Config) ->
-    case lists:filter(fun(#config{key = hosts}) -> true;
-                         (_) -> false
-                      end, Config) of
+-spec get_key(config_list(), mongoose_config_parser:key()) ->
+    [mongoose_config_parser:value()].
+get_key(Config, Key) ->
+    FilterFn = fun(#config{key = K}) when K =:= Key -> true;
+                  (_) -> false
+               end,
+    case lists:filter(FilterFn, Config) of
         [] -> [];
-        [#config{value = Hosts}] -> Hosts
+        [#config{value = Value}] -> Value
     end.
 
--spec build_state([jid:server()], [top_level_config()], [override()]) ->
+-spec build_state([jid:server()], [jid:server()], [top_level_config()], [override()]) ->
           mongoose_config_parser:state().
-build_state(Hosts, Opts, Overrides) ->
+build_state(Hosts, HostTypes, Opts, Overrides) ->
     lists:foldl(fun(F, StateIn) -> F(StateIn) end,
                 mongoose_config_parser:new_state(),
                 [fun(S) -> mongoose_config_parser:set_hosts(Hosts, S) end,
+                 fun(S) -> mongoose_config_parser:set_host_types(HostTypes, S) end,
                  fun(S) -> mongoose_config_parser:set_opts(Opts, S) end,
                  fun mongoose_config_parser:dedup_state_opts/1,
                  fun mongoose_config_parser:add_dep_modules/1,
                  fun(S) -> set_overrides(Overrides, S) end]).
 
+-spec check_dynamic_domains_support(mongoose_config_parser:state()) -> any().
```

Comment:
not sure if `src/config/mongoose_config_parser_toml.erl` is the right place for this check, MB `src/config/mongoose_config_parser.erl` fits better.
@chrzaszcz, what's your opinion on it?

Reply:
If possible, I'd rather have it in `src/config/mongoose_config_parser.erl`